controllers/operatorconfig_controller.go

```diff
@@ -54,51 +78,326 @@ type OperatorConfigReconciler struct {
+		}
 	} else {
-		log.Info("Proxy Configured")
+		log.Info(fmt.Sprintf("Created %s/%s", obj.Kind, obj.Name))
 	}
 
-	return ctrl.Result{}, err
+	return nil
 }
 
-func configureProxy(ctx context.Context, log logr.Logger) error {
-	return createResourcesFromFile("proxy.yaml", ctx, log)
+func (r *OperatorConfigReconciler) reconcileClusterTask(resource *unstructured.Unstructured, ctx context.Context, log logr.Logger, oc *cranev1alpha1.OperatorConfig) error {
+	var obj pipelinev1beta1.ClusterTask
+	err := runtime.DefaultUnstructuredConverter.
+		FromUnstructured(resource.UnstructuredContent(), &obj)
+	if err != nil {
+		return err
+	}
+
+	// Add Finalizer for this CR
+	if !controllerutil.ContainsFinalizer(&obj, Finalizer) {
+		controllerutil.AddFinalizer(&obj, Finalizer)
+	}
+
+	clusterTaskLabel := obj.Labels
+	if clusterTaskLabel != nil {
+		clusterTaskLabel[OwnerLabel] = OwnerConfigName
+	} else {
+		clusterTaskLabel = map[string]string{
+			OwnerLabel: OwnerConfigName,
+		}
+	}
+	obj.Labels = clusterTaskLabel
+
+	if err := r.Create(ctx, &obj); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			log.Error(err, fmt.Sprintf("Failed to create %s", obj.Kind), "Namespace", obj.Namespace, "Name", obj.Name)
+			return err
+		}
+	} else {
+		log.Info(fmt.Sprintf("Created %s/%s", obj.Kind, obj.Name))
+	}
+
+	c := pipelinev1beta1.ClusterTask{}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: obj.Namespace, Name: obj.Name}, &c); err != nil {
+		return err
+	}
+
+	if !reflect.DeepEqual(c.Spec, obj.Spec) {
+		c.Spec = obj.Spec
+		if err := r.Update(ctx, &c); err != nil {
+			log.Error(err, fmt.Sprintf("Failed to update %s", obj.Kind), "Namespace", obj.Namespace, "Name", obj.Name)
+			return err
+		}
+	}
+
+	return nil
 }
 
-func configureCranePlugin(ctx context.Context, log logr.Logger) error {
-	return createResourcesFromFile("crane-ui-plugin.yaml", ctx, log)
+func (r *OperatorConfigReconciler) reconcileConsolePlugin(resource *unstructured.Unstructured, ctx context.Context, log logr.Logger, oc *cranev1alpha1.OperatorConfig) error {
+	var obj consolev1alpha1.ConsolePlugin
+	err := runtime.DefaultUnstructuredConverter.
+		FromUnstructured(resource.UnstructuredContent(), &obj)
+	if err != nil {
+		return err
+	}
+
+	consolePluginLabel := obj.Labels
+	if consolePluginLabel != nil {
+		consolePluginLabel[OwnerLabel] = OwnerConfigName
+	} else {
+		consolePluginLabel = map[string]string{
+			OwnerLabel: OwnerConfigName,
+		}
+	}
+	obj.Labels = consolePluginLabel
+
+	if err := r.Create(ctx, &obj); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			log.Error(err, fmt.Sprintf("Failed to create %s", obj.Kind), "Namespace", obj.Namespace, "Name", obj.Name)
+			return err
+		}
+	} else {
+		log.Info(fmt.Sprintf("Created %s/%s", obj.Kind, obj.Name))
+	}
+
+	c := consolev1alpha1.ConsolePlugin{}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: obj.Namespace, Name: obj.Name}, &c); err != nil {
+		return err
+	}
+
+	if !reflect.DeepEqual(c.Spec, obj.Spec) {
```

Comment:
on top of the spec, we also need to compare labels? right?


I re-commend using `CreateOrUpdate` method of controller runtime to squeeze the create and updates in one step 

https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrUpdate there is an example here 

and oadp-operator has been using this a lot, https://github.com/openshift/oadp-operator/blob/89d5bccad2ed23509f0cf6ab49b6f694dd298591/controllers/velero.go#L259, I think it will save you a lot of trouble

Reply:
Thanks for this, helped a lot!